saleor/core/taxes/avatax/__init__.py

```diff
@@ -0,0 +1,388 @@
+        return False
+    shipping_address = order.shipping_address
+    is_shipping_required = order.is_shipping_required()
+    address = shipping_address or order.billing_address
+
+    if not is_shipping_required and not address:
+        return False
+    if not shipping_address:
+        return False
+    if not order.shipping_method:
+        return False
+    return True
+
+
+def validate_checkout(checkout: "Checkout") -> bool:
+    """Validate if checkout object contains enough information to generate a request to
+    avatax"""
+    if not checkout.lines.count():
+        return False
+
+    shipping_address = checkout.shipping_address
+    is_shipping_required = checkout.is_shipping_required
+    address = shipping_address or checkout.billing_address
+    if not is_shipping_required and not address:
+        return False
+    if not shipping_address:
+        return False
+    if not checkout.shipping_method:
+        return False
+    return True
+
+
+def checkout_needs_new_fetch(data, checkout_token: str) -> bool:
+    """We store the response from avatax for checkout object for given time. If object
+    doesn't exist in cache or something has changed, then we fetch data from avatax."""
+    checkout_cache_key = settings.AVATAX_CACHE_KEY + checkout_token
+    cached_checkout = cache.get(checkout_cache_key)
+
+    if not cached_checkout:
+        return True
+
+    cached_request_data, cached_response = cached_checkout
+    if data != cached_request_data:
+        return True
+    return False
+
+
+def taxes_need_new_fetch(data: Dict[str, Any], taxes_token: str) -> bool:
+    """We store the response from avatax. If object doesn't exist in cache or
+    something has changed, then we fetch data from avatax."""
+    taxes_cache_key = settings.AVATAX_CACHE_KEY + taxes_token
+    cached_data = cache.get(taxes_cache_key)
+
+    if not cached_data:
+        return True
+
+    cached_request_data, _ = cached_data
+    if data != cached_request_data:
+        return True
+    return False
+
+
+def append_line_to_data(
+    data: List[Dict[str, str]],
+    quantity: int,
+    amount: str,
+    tax_code: str,
+    item_code: str,
+    description: str = None,
+):
+    data.append(
+        {
+            "quantity": quantity,
+            "amount": str(amount),
+            "taxCode": tax_code,
+            "taxIncluded": include_taxes_in_prices(),
+            "itemCode": item_code,
+            "description": description[:2000] if description else "",
+        }
+    )
+
+
+def get_checkout_lines_data(
+    checkout: "Checkout", discounts=None
+) -> List[Dict[str, str]]:
+    data = []
+    lines = checkout.lines.prefetch_related(
+        "variant__product__category",
+        "variant__product__collections",
+        "variant__product__product_type",
+    )
+    for line in lines:
+        if not line.variant.product.charge_taxes:
+            continue
+        description = line.variant.product.description
```

Comment:
That could generate huge payloads. Is this really needed/useful?

Reply:
It can be helpful on Avalara's admin panel. `description` is passing to `append_line_to_data` and `append_line_to_data ` has limit to 2k characters:
`"description": description[:2000] if description else "",`
If this number is too big I will change it to lower.